### PR TITLE
feat(capacity-tx): make fees stable

### LINKED ISSUE #1015: [Capacity] Ensure the stability of Capacity fees (Capacity tx fee)
As the network that wants to ensure the stability of Capacity, the fee for a Capacity transaction is based on [something].

AC:
- Transaction cost should not change when generating new weights for Capacity transactions.
- The cost of a Capacity transaction stays the same regardless of the newly generated weights.
- The generated fee should not consider network congestion for Capacity.

### PR DESCRIPTION
**Pull request by enddynayn**
# Goal
Make capacity fees stable to weight changes and
network congestion.

By making capacity fees stable, users have the
ability to always know how many transactions they can
send during an epoch period.

Unlike token transactions, capacity transactions
do not allow tipping or multiplying the weight
fee by a targeted_fee_adjustment.

The fee for a capacity transaction:
> fee = base_fee + length_fee + weight_fee

Note that weights for capacity are a
snapshot of weights calculated from benchmarks.
These weights are static and are not updated.

Closes #1015

### CODE REVIEW

File: pallets/frequency-tx-payment/src/capacity_payment.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+
+use pallet_messages::Call as MessagesCall;
```
Comment (enddynayn) [inline pallets/frequency-tx-payment/src/capacity_payment.rs:2]: remove

File: pallets/frequency-tx-payment/src/lib.rs [hunk 2]
```diff
@@ -180,6 +189,22 @@ pub mod pallet {
...
+impl<T: Config> Pallet<T> {
+	pub fn compute_capacity_fee(
+		len: u32,
+		call: &<T as Config>::RuntimeCall,
+	) -> Option<BalanceOf<T>> {
+		let weight = <T as Config>::CapacityCalls::get_stable_weight(call)?;
+		let len_fee = T::LengthToFee::weight_to_fee(&Weight::from_ref_time(len as u64));
+
+		let capped_weight = weight.min(T::BlockWeights::get().max_block);
+		let base_fee = T::WeightToFee::weight_to_fee(&capped_weight);
```
Comment (enddynayn) [inline pallets/frequency-tx-payment/src/lib.rs:201]: Make this a constant?

**Comment by enddynayn** [comment]
Any suggestions on how to improve the design of this are welcome!

File: pallets/frequency-tx-payment/src/capacity_stable_weights.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+			.saturating_add(T::DbWeight::get().reads(6 as u64))
+			.saturating_add(T::DbWeight::get().writes(5 as u64))
+	}
+	// Storage: Msa PayloadSignatureBucketCount (r:1 w:1)
+	// Storage: Msa PayloadSignatureRegistry (r:1 w:1)
+	// Storage: Msa PublicKeyToMsaId (r:2 w:0)
+	// Storage: Msa ProviderToRegistryEntry (r:1 w:0)
+	// Storage: Msa DelegatorAndProviderToDelegation (r:1 w:1)
+	// Storage: Schemas CurrentSchemaIdentifierMaximum (r:1 w:0)
+	fn grant_delegation(s: u32, ) -> Weight {
```
Comment (enddynayn) [inline pallets/frequency-tx-payment/src/capacity_stable_weights.rs:99]: @wilwade can you please check these to see if we are missing something?

File: pallets/frequency-tx-payment/src/lib.rs [hunk 4]
```diff
@@ -188,6 +213,8 @@ pub enum ChargeFrqTransactionPaymentError {
 	InvalidMsaKey,
 	/// The Capacity Target does not exist
 	TargetCapacityNotFound,
+	/// The call has not no stable weights associated with it.
```
Comment (shannonwells) [inline pallets/frequency-tx-payment/src/lib.rs:216]:
```suggestion
	/// The call has no stable weights associated with it.
```

**Review by shannonwells — COMMENTED** [review]
I'm a bit concerned about duplicated capacity call lists. I was imaginging one entity, maybe an enum of calls and an implementation of Contains and GetStableWeight on the enum.  I don't know what you've tried already but maybe we can pair on this.

**Comment by enddynayn** [comment]
> I'm a bit concerned about duplicated capacity call lists. I was imaginging one entity, maybe an enum of calls and an implementation of Contains and GetStableWeight on the enum.  I don't know what you've tried already but maybe we can pair on this.

I totally agree. I tried the enum thing as you suggested but could not get it to work.

File: pallets/frequency-tx-payment/src/lib.rs [hunk 5]
```diff
@@ -180,6 +189,37 @@ pub mod pallet {
...
+	pub fn compute_capacity_fee(
+		len: u32,
+		class: DispatchClass,
+		call: &<T as Config>::RuntimeCall,
+	) -> Option<BalanceOf<T>> {
+		let weight_fee =
+			Self::weight_to_fee(<T as Config>::CapacityCalls::get_stable_weight(call)?);
+
+		let len_fee = Self::length_to_fee(len);
+		let base_fee = Self::weight_to_fee(T::BlockWeights::get().get(class).base_extrinsic);
```
Comment (enddynayn) [inline pallets/frequency-tx-payment/src/lib.rs:202]: Maybe update to a constant.

File: pallets/frequency-tx-payment/src/lib.rs [hunk 6]
```diff
@@ -148,11 +151,14 @@ pub mod pallet {
...
-		type Capacity: Nontransferable + Replenishable;
+		/// The type that replenishes and keeps capacity balances.
+		type Capacity: Replenishable + Nontransferable;
 
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
-		type CapacityEligibleCalls: Contains<<Self as Config>::RuntimeCall>;
+
+		/// The type that checks what transactions are capacity with their stable weights.
+		type CapacityCalls: GetStableWeight<<Self as Config>::RuntimeCall, Weight>;
```
Comment (shannonwells) [inline pallets/frequency-tx-payment/src/lib.rs:161]: ❤️

**Review by shannonwells — CHANGES_REQUESTED** [review]
I think converting Contains trait to GetStableWeight was a great change.
The only blocking comment I have is about the Error name.

File: pallets/frequency-tx-payment/src/lib.rs [hunk 7]
```diff
@@ -188,6 +224,8 @@ pub enum ChargeFrqTransactionPaymentError {
 	InvalidMsaKey,
 	/// The Capacity Target does not exist
 	TargetCapacityNotFound,
+	/// The call has no stable weights associated with it.
+	CallHasNoStableWeights,
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/frequency-tx-payment/src/lib.rs:228]: ahh good catch.

File: pallets/frequency-tx-payment/src/lib.rs (hunk 5 above)
Comment (enddynayn) [reply pallets/frequency-tx-payment/src/lib.rs:198]: Ah yeah, I am just waiting for @wilwade on this one.

File: pallets/frequency-tx-payment/src/lib.rs [hunk 8]
```diff
@@ -237,16 +275,10 @@ where
...
-			pallet_transaction_payment::Pallet::<T>::compute_fee(len as u32, info, self.tip(call));
-
 		match call.is_sub_type() {
 			Some(Call::pay_with_capacity { call }) => {
-				use frame_support::traits::Contains;
-				ensure!(
-					<T as Config>::CapacityEligibleCalls::contains(call.as_ref()),
-					ChargeFrqTransactionPaymentError::CallIsNotCapacityEligible.into()
-				);
+				let extrinsic_weight = T::CapacityCalls::get_stable_weight(call)
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/frequency-tx-payment/src/lib.rs:278]: Me too. Feel free to let me know if you have any good ideas here. I thought about it here but could not find a good pattern and name.

File: pallets/frequency-tx-payment/src/lib.rs [hunk 9]
```diff
@@ -281,10 +275,8 @@ where
 	) -> Result<(BalanceOf<T>, InitialPayment<T>), TransactionValidityError> {
 		match call.is_sub_type() {
 			Some(Call::pay_with_capacity { call }) => {
-				ensure!(
-					T::CapacityCalls::contains(call.as_ref()),
-					ChargeFrqTransactionPaymentError::CallIsNotCapacityEligible.into()
-				);
+				let extrinsic_weight = T::CapacityCalls::get_stable_weight(call)
```
Comment (shannonwells) [inline pallets/frequency-tx-payment/src/lib.rs:278]: better

**Review by shannonwells — APPROVED** [review]

File: pallets/frequency-tx-payment/src/lib.rs (hunk 9 above)
Comment (enddynayn) [reply pallets/frequency-tx-payment/src/lib.rs:278]: Nothing changed 😄.

File: pallets/frequency-tx-payment/src/lib.rs [hunk 10]
```diff
@@ -344,7 +389,7 @@ where
 		info: &DispatchInfoOf<Self::Call>,
 		len: usize,
 	) -> Result<Self::Pre, TransactionValidityError> {
-		let (_fee, initial_payment) = self.withdraw_fee(who, call, info, len)?;
+		let (_fee, initial_payment) = self.withdraw_fee(who, &call, info, len)?;
```
[the remark this answers is not in the document]
Comment (johnmela) [reply pallets/frequency-tx-payment/src/lib.rs:390]: I like this as well

**Review by johnmela — APPROVED** [review]
Approved 👍 

It would be a nice to have to get a sanity check in the form of a test that ensures that white listed call weight is stable. Non blocking of course.

File: pallets/frequency-tx-payment/src/lib.rs (hunk 5 above)
Comment (enddynayn) [reply pallets/frequency-tx-payment/src/lib.rs:198]: I am going to create a ticket to update if we decide to do so.

**Comment by enddynayn** [comment]
> Approved 👍
> 
> It would be a "nice to have" to get a sanity check in the form of a test that ensures that white listed call fees are stable. Non blocking of course.

The test is there, but maybe I misunderstood the test you are thinking about.

**Pull request merged**